libnd4j/include/helpers/impl/AttentionHelper.cpp

```diff
@@ -0,0 +1,95 @@
+/*******************************************************************************
+ * Copyright (c) 2015-2019 Skymind, Inc.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License, Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ******************************************************************************/
+
+//
+// @author Paul Dubs
+//
+
+#ifndef LIBND4J_ATTENTIONHELPER_CPP
+#define LIBND4J_ATTENTIONHELPER_CPP
+
+#include <helpers/AttentionHelper.h>
+
+#include "../AttentionHelper.h"
+#include <ops/declarable/CustomOperations.h>
+
+namespace nd4j {
+
+    nd4j::NDArray *
+    AttentionHelper::multiHeadProject(const nd4j::NDArray *input, const nd4j::NDArray *projectionMatrix) {
+        auto miniBatchSize = input->sizeAt(0);
+        auto seqLength = input->sizeAt(2);
+        auto numHeads = projectionMatrix->sizeAt(0);
+        auto projectedSize = projectionMatrix->sizeAt(1);
+
+        auto inputPerm = input->permute({1, 0, 2});
+        auto inputPrep = inputPerm->reshape(input->ordering(), {input->sizeAt(1), (miniBatchSize * seqLength)});
```

Comment:
With this permute - what happens if I call the method with an array with weird strides?
Probably best to add a few extra tests (or grad checks) for this case. Not sure if the libnd4j permute+reshape matches the Nd4J behaviour... but if it does: this could be a problem.